### DISCUSSION TITLE
Why does `Marshal.SizeOf<T>()` require non-generic types, but `Marshal.SizeOf<T>(T)` doesn't?

### DISCUSSION THREAD

**Discussion by colejohnson66** [dotnet/runtime#96487] [Q&A] [answered]
I wanted to use `Marshal.SizeOf<T>()` to get the size of an object, but it apparently throws an `ArgumentException` (undocumented if I may add) if `T` is a generic type:

https://github.com/dotnet/runtime/blob/0d8fadd87774dda07cf943056161204e4fe8b981/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs#L122-L131

However, curiously, the overload that takes an object of type `T` (`Marshal.SizeOf<T>(T)`) *doesn't* throw for generic types:

https://github.com/dotnet/runtime/blob/0d8fadd87774dda07cf943056161204e4fe8b981/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs#L97-L102

A simple example using nullable value types (`Nullable<T>`):

```csharp
using System;
using System.Runtime.InteropServices;

Console.WriteLine(Marshal.SizeOf<int?>());     // throws `ArgumentException`
Console.WriteLine(Marhsal.SizeOf<int?>(42));   // prints "4" (not "8" as `sizeof(int?)` would, but ignore that)
Console.WriteLine(Marshal.SizeOf<int?>(new()); // ditto
```

Is there a reason for this restriction?

**Comment by tannergooding** [comment, marked as answer]
`Marshal.SizeOf<T>()` is intended to get the size of the object after it's been marshalled and so is effectively the number of bytes that native sees.

`sizeof(T)` (and `Unsafe.SizeOf<T>()` is the size of the object in managed, which will match the size for blittable types.

Historically, generics were not supported in marshalling and so could not be passed across the P/Invoke boundary, this was changed around .NET Core 3.1, but many of the `Marshal` APIs were not updated to account for that as the built-in marshalling was already being considered "legacy" and was being looked at getting replaced (see https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.disableruntimemarshallingattribute, https://learn.microsoft.com/en-us/dotnet/standard/native-interop/pinvoke-source-generation, etc)

`Nullable<T>` is additionally a special type and doesn't always behave like a normal generic, has specialized type system support/handling, and doesn't have a clear 1-to-1 with any native types. It is explicitly blocked from use in marshalling scenarios and so is not supported: https://github.com/dotnet/runtime/blob/04afbf73724a255078e78bfd710418e339627b2a/src/coreclr/vm/mlinfo.cpp#L1084-L1101

CC. @AaronRobinsonMSFT, @jkoritzinsky who may be able to provide additional context/info

**Reply by AaronRobinsonMSFT** [reply]
I don't have much to add to @tannergooding's comprehensive response above. As mentioned the `Marshal.SizeOf<T>()` series of APIs don't really do what most people want them to do and if they really do need them they are expensive in general and not the best way to address interop scenarios.

> I wanted to use Marshal.SizeOf<T>() to get the size of an object

I would look at https://github.com/dotnet/runtime/issues/24200 for more details. The interop scenario is complicated for instances of type `object` and that is what the `Marshal` APIs are for, interop. That means that size queries on arbitrary `object`s are not what is expected. For scenarios involving allocation schemes of managed `object`s the above issue is a more appropriate place to look for guidance and where we are likely to invest.

**Reply by colejohnson66** [reply]
So the answer is that the limitation is just a historical accident, not necessarily a valid one?

**Comment by huoyaoyuan** [comment]
I'm not sure about the limitation, but `Marshal.SizeOf` isn't the size of *the object*. It's the size of native counterpart under effective marshaling rule. Use `Unsafe.SizeOf` or `sizeof` operator for managed object size.
